tools/klee/main.cpp

```diff
@@ -469,53 +496,120 @@ KleeHandler::openTestFile(const std::string &suffix, unsigned id) {
+    *file << item.first;
+    *file << "\">";
+    // Ignore the type
+    auto type_size_bytes = item.second.size() * 8;
+    llvm::APInt v(type_size_bytes, 0, false);
+    for (auto i = item.second.rbegin(), e = item.second.rend(); i != e; ++i) {
+      v <<= 8;
+      v |= *i;
+    }
+    // print value
+
+    // Check if this is an unsigned type
+    if (item.first.find("u") == 0) {
+      v.print(*file, false);
+    } else if (item.first.rfind("*") != std::string::npos) {
+      // Pointer types
+      v.print(*file, false);
+    } else if (item.first.find("float") == 0) {
+      llvm::APFloat(APFloatBase::IEEEhalf(), v).print(*file);
+    } else if (item.first.find("double") == 0) {
+      llvm::APFloat(APFloatBase::IEEEdouble(), v).print(*file);
+    } else if (item.first.rfind("_t") != std::string::npos) {
+      // arbitrary type, e.g. sector_t
+      v.print(*file, false);
+    } else if (item.first.find("_") == 0) {
+      // _Bool
+      v.print(*file, false);
+    } else {
+      // the rest must be signed
+      v.print(*file, true);
+    }
+    *file << "</input>\n";
+  }
+  *file << "</testcase>\n";
 
+  ++m_numGeneratedTests;
+}
 /* Outputs all files (.ktest, .kquery, .cov etc.) describing a test case */
 void KleeHandler::processTestCase(const ExecutionState &state,
                                   const char *errorMessage,
                                   const char *errorSuffix) {
   if (!WriteNone) {
-    std::vector< std::pair<std::string, std::vector<unsigned char> > > out;
-    bool success = m_interpreter->getSymbolicSolution(state, out);
+    const auto start_time = time::getWallTime();
+    std::vector< std::pair<std::string, std::vector<unsigned char> > > assignments;
+    bool success = m_interpreter->getSymbolicSolution(state, assignments);
 
     if (!success)
       klee_warning("unable to get symbolic solution, losing test case");
 
-    const auto start_time = time::getWallTime();
-
-    unsigned id = ++m_numTotalTests;
+    unsigned test_id = ++m_numTotalTests;
 
     if (success) {
-      KTest b;
-      b.numArgs = m_argc;
-      b.args = m_argv;
-      b.symArgvs = 0;
-      b.symArgvLen = 0;
-      b.numObjects = out.size();
-      b.objects = new KTestObject[b.numObjects];
-      assert(b.objects);
-      for (unsigned i=0; i<b.numObjects; i++) {
-        KTestObject *o = &b.objects[i];
-        o->name = const_cast<char*>(out[i].first.c_str());
-        o->numBytes = out[i].second.size();
-        o->bytes = new unsigned char[o->numBytes];
-        assert(o->bytes);
-        std::copy(out[i].second.begin(), out[i].second.end(), o->bytes);
-      }
-
-      if (!kTest_toFile(&b, getOutputFilename(getTestFilename("ktest", id)).c_str())) {
-        klee_warning("unable to write output test case, losing it");
-      } else {
-        ++m_numGeneratedTests;
-      }
-
-      for (unsigned i=0; i<b.numObjects; i++)
-        delete[] b.objects[i].bytes;
-      delete[] b.objects;
+      if (!WriteXMLTests)
```

Comment:
I don't think this should be an OR, we should allow the generation of both .ktest and .xml files